tests/test_client_middleware.py

```diff
@@ -847,11 +847,12 @@ async def __call__(
             retry_count = 0
             while True:
                 self.attempt_count += 1
+                response = await handler(request)
                 if retry_count == 0:
                     retry_count += 1
-                    await handler(request)
+                    response.release()  # Release the response to enable connection reuse
```

Comment:
I was just thinking through this one. Shouldn't the connection already have been released if the body was fully read? Given the body is only a couple of bytes, it should have finished reading the response by this point in order to have the headers set.